File: services/api-gateway/lib/rateLimiter.js
const TOKEN_BUCKET_SCRIPT = `
local key = KEYS[1]
local capacity = tonumber(ARGV[1])
local refill_rate = tonumber(ARGV[2])
local now = tonumber(ARGV[3])
local requested = tonumber(ARGV[4])
local tokens = tonumber(redis.call("HGET", key, "tokens"))
local ts = tonumber(redis.call("HGET", key, "ts"))
if tokens == nil then tokens = capacity end
if ts == nil then ts = now end
local elapsed = now - ts
if elapsed < 0 then elapsed = 0 end
tokens = math.min(capacity, tokens + elapsed * refill_rate)
local allowed = 0
if tokens >= requested then
  allowed = 1
  tokens = tokens - requested
end
redis.call("HSET", key, "tokens", tokens, "ts", now)
redis.call("EXPIRE", key, 3600)
return allowed
`;

class TokenBucketRateLimiter {
  constructor(redisClient, config) {
    this.redis = redisClient;
    this.config = config;
    this.localState = new Map();
    this.rejections = new Map();
  }

  refill(tokens, ts, now) {
    const elapsed = Math.max(0, now - ts);
    return Math.min(this.config.capacity, tokens + elapsed * this.config.refillRate);
  }

  async allowRedis(identifier, tokensRequested) {
    const key = `ratelimit:edge:${identifier}`;
    const now = Date.now() / 1000;
    const result = await this.redis.eval(
      TOKEN_BUCKET_SCRIPT,
      1,
      key,
      this.config.capacity,
      this.config.refillRate,
      now,
      tokensRequested
    );
    return result === 1;
  }

  async allow(identifier, tokensRequested = 1) {
    let allowed;
    if (this.redis) {
      allowed = await this.allowRedis(identifier, tokensRequested);
    } else {
      const now = Date.now() / 1000;
      const state = this.localState.get(identifier) || { tokens: this.config.capacity, ts: now };
      state.tokens = this.refill(state.tokens, state.ts, now);
      state.ts = now;
      if (state.tokens >= tokensRequested) {
        state.tokens -= tokensRequested;
        allowed = true;
      } else {
        allowed = false;
      }
      this.localState.set(identifier, state);
    }
    if (!allowed) {
      this.rejections.set(identifier, (this.rejections.get(identifier) || 0) + 1);
    }
    return allowed;
  }

  rejectionCounts() {
    return Object.fromEntries(this.rejections);
  }
}

module.exports = { TokenBucketRateLimiter };

File: shared/shared/rate_limiter.py
"""Token bucket rate limiter. Atomicity on Redis is done via optimistic
locking (WATCH/MULTI/EXEC) instead of a Lua script: WATCH the bucket key,
read+compute the new token count client-side, then commit with MULTI/EXEC.
If another client wrote to the key in between, EXEC aborts (WatchError) and
we retry read-compute-commit — same end result as a server-side Lua script,
just via optimistic concurrency + retry instead of running our logic inside
Redis. Trade-off: a few round-trips per call instead of one eval, and a
retry loop under heavy contention on the same key — acceptable here since
each agent type has its own bucket key, so contention is per-agent-type, not
global. Falls back to a local in-memory bucket (same math) when no Redis
client is supplied, so the module is testable without a Redis server.
"""
import time
from dataclasses import dataclass
from typing import Optional


@dataclass
class BucketConfig:
    capacity: float
    refill_rate: float  # tokens per second


class TokenBucketRateLimiter:
    """Per-agent-type token buckets. Uses Redis WATCH/MULTI/EXEC if a client
    is supplied, otherwise an in-process dict (identical refill math)."""

    def __init__(self, redis_client=None, configs: Optional[dict] = None, max_retries: int = 5):
        self.redis = redis_client
        self.max_retries = max_retries
        self.configs = configs or {
            "text_agent": BucketConfig(capacity=20, refill_rate=5),
            "image_agent": BucketConfig(capacity=8, refill_rate=1.5),
            "table_agent": BucketConfig(capacity=15, refill_rate=4),
            "synthesizer": BucketConfig(capacity=10, refill_rate=2),
        }
        self._local_state: dict = {}
        self._rejections: dict = {}  # agent_type -> count, real counter (was a hardcoded {} in /status before)

    def _refill(self, tokens: float, ts: float, now: float, cfg: BucketConfig) -> float:
        elapsed = max(0.0, now - ts)
        return min(cfg.capacity, tokens + elapsed * cfg.refill_rate)

    def _allow_redis(self, agent_type: str, tokens_requested: float, cfg: BucketConfig) -> bool:
        key = f"ratelimit:{agent_type}"
        for _ in range(self.max_retries):
            pipe = self.redis.pipeline()
            try:
                pipe.watch(key)
                raw = pipe.hmget(key, "tokens", "ts")
                tokens = float(raw[0]) if raw[0] is not None else cfg.capacity
                ts = float(raw[1]) if raw[1] is not None else time.time()

                now = time.time()
                tokens = self._refill(tokens, ts, now, cfg)
                allowed = tokens >= tokens_requested
                if allowed:
                    tokens -= tokens_requested

                pipe.multi()
                pipe.hset(key, mapping={"tokens": tokens, "ts": now})
                pipe.expire(key, 3600)
                pipe.execute()  # raises WatchError if the key changed since WATCH
                return allowed
            except Exception as e:
                # redis-py raises WatchError (subclass of Exception) on a
                # concurrent write; retry read-compute-commit from scratch.
                import redis
                if not isinstance(e, redis.exceptions.WatchError):
                    raise
                continue
            finally:
                pipe.reset()
        # exhausted retries under heavy contention on this key: fail closed
        return False

    def allow(self, agent_type: str, tokens_requested: float = 1) -> bool:
        cfg = self.configs[agent_type]
        if self.redis is not None:
            allowed = self._allow_redis(agent_type, tokens_requested, cfg)
        else:
            # in-memory fallback, same refill formula as the Redis path
            now = time.time()
            state = self._local_state.setdefault(agent_type, {"tokens": cfg.capacity, "ts": now})
            state["tokens"] = self._refill(state["tokens"], state["ts"], now, cfg)
            state["ts"] = now
            if state["tokens"] >= tokens_requested:
                state["tokens"] -= tokens_requested
                allowed = True
            else:
                allowed = False

        if not allowed:
            self._rejections[agent_type] = self._rejections.get(agent_type, 0) + 1
        return allowed

    def rejection_counts(self) -> dict:
        """Real per-agent-type rejection counts since process start. Process-
        local even in Redis mode (the token state is shared across replicas,
        this counter isn't) -- fine for a single-instance dashboard stat."""
        return dict(self._rejections)
